File: frontend/src/lib/readonly.test.ts
import { describe, it, expect } from "vitest";
import { can, isWriteBlocked } from "./readonly";

// The two modes the app ships in: a desktop editor and a small-screen viewer.
const EDITOR = false;
const VIEWER = true;

describe("can", () => {
  it("withholds authoring from a viewer", () => {
    expect(can("workflow.create", VIEWER)).toBe(false);
    expect(can("workflow.delete", VIEWER)).toBe(false);
    expect(can("workflow.editGraph", VIEWER)).toBe(false);
    expect(can("workflow.deploy", VIEWER)).toBe(false);
    expect(can("workflow.buildFromChat", VIEWER)).toBe(false);
  });

  it("lets a viewer operate a workflow somebody else built", () => {
    expect(can("workflow.run", VIEWER)).toBe(true);
    expect(can("workflow.stop", VIEWER)).toBe(true);
    expect(can("workflow.chat", VIEWER)).toBe(true);
  });

  it("leaves a viewer's account alone", () => {
    expect(can("account.billing", VIEWER)).toBe(true);
    expect(can("account.settings", VIEWER)).toBe(true);
  });

  // The whole point of the change: a desktop client is not restricted at all.
  it("withholds nothing from an editor", () => {
    const every = [
      "workflow.create",
      "workflow.delete",
      "workflow.editGraph",
      "workflow.deploy",
      "workflow.buildFromChat",
      "workflow.run",
      "workflow.stop",
      "workflow.chat",
      "account.billing",
      "account.settings",
    ] as const;
    for (const c of every) expect(can(c, EDITOR)).toBe(true);
  });
});

describe("isWriteBlocked", () => {
  it("blocks the five graph-mutating endpoints for a viewer", () => {
    expect(isWriteBlocked("POST", "/workflows", VIEWER)).toBe(true);
    expect(isWriteBlocked("PUT", "/workflows/wf_123", VIEWER)).toBe(true);
    expect(isWriteBlocked("DELETE", "/workflows/wf_123", VIEWER)).toBe(true);
    expect(isWriteBlocked("POST", "/workflows/wf_123/deploy", VIEWER)).toBe(
      true,
    );
    expect(isWriteBlocked("POST", "/workflows/wf_123/build", VIEWER)).toBe(
      true,
    );
  });

  // Added alongside PUT/DELETE .../schedule, PUT/DELETE .../geofence, and
  // GET /tendril/console -- these must mirror backend/internal/api/readonly.go
  // one for one, same as every other rule in this list.
  it("blocks schedule, geofence, and the tendril console for a viewer", () => {
    expect(isWriteBlocked("PUT", "/workflows/wf_123/schedule", VIEWER)).toBe(
      true,
    );
    expect(
      isWriteBlocked("DELETE", "/workflows/wf_123/schedule", VIEWER),
    ).toBe(true);
    expect(isWriteBlocked("PUT", "/workflows/wf_123/geofence", VIEWER)).toBe(
      true,
    );
    expect(
      isWriteBlocked("DELETE", "/workflows/wf_123/geofence", VIEWER),
    ).toBe(true);
    // GET, not a write verb -- listed because the backend handler behind it
    // creates a workflow row on first call (get-or-create), so it is a write
    // in effect. See lib/tendril.ts's console().
    expect(isWriteBlocked("GET", "/tendril/console", VIEWER)).toBe(true);
    // The Prism console's pair behaves identically. /prism/run stays open
    // because running a paid call is operating, not authoring -- the line
    // /tendril/run and /tendril/topup already sit on.
    expect(isWriteBlocked("GET", "/prism/console", VIEWER)).toBe(true);
    expect(isWriteBlocked("GET", "/prism/console/exists", VIEWER)).toBe(false);
    expect(isWriteBlocked("POST", "/prism/run", VIEWER)).toBe(false);
    expect(isWriteBlocked("GET", "/prism/endpoints", VIEWER)).toBe(false);
  });

  it("blocks nothing for an editor", () => {
    expect(isWriteBlocked("POST", "/workflows", EDITOR)).toBe(false);
    expect(isWriteBlocked("DELETE", "/workflows/wf_123", EDITOR)).toBe(false);
    expect(isWriteBlocked("POST", "/workflows/wf_1/deploy", EDITOR)).toBe(
      false,
    );
  });

  it("lets a run, a stop, and every read through", () => {
    expect(isWriteBlocked("POST", "/workflows/wf_123/run", VIEWER)).toBe(false);
    expect(isWriteBlocked("POST", "/workflows/wf_123/stop", VIEWER)).toBe(
      false,
    );
    expect(isWriteBlocked("GET", "/workflows", VIEWER)).toBe(false);
    expect(isWriteBlocked("GET", "/workflows/wf_123", VIEWER)).toBe(false);
    expect(isWriteBlocked("POST", "/credits/redeem-coupon", VIEWER)).toBe(
      false,
    );
    expect(isWriteBlocked("PATCH", "/settings", VIEWER)).toBe(false);
    expect(isWriteBlocked("POST", "/auth/password", VIEWER)).toBe(false);
  });

  it("is case-insensitive on the method", () => {
    expect(isWriteBlocked("post", "/workflows", VIEWER)).toBe(true);
    expect(isWriteBlocked("delete", "/workflows/wf_1", VIEWER)).toBe(true);
  });

  // A blocked path must stay blocked however it is dressed up, or the guard is
  // decorative -- these are the shapes a fetch helper can produce by accident.
  it("cannot be slipped past with a query string or trailing slash", () => {
    expect(isWriteBlocked("POST", "/workflows?draft=1", VIEWER)).toBe(true);
    expect(isWriteBlocked("POST", "/workflows/", VIEWER)).toBe(true);
    expect(isWriteBlocked("PUT", "/workflows/wf_1/", VIEWER)).toBe(true);
    expect(
      isWriteBlocked("POST", "/workflows/wf_1/deploy?force=1", VIEWER),
    ).toBe(true);
  });

  // The id segment is one path element. A deeper path is a different endpoint
  // and must not be swept up by the single-segment rules.
  it("does not over-match nested routes", () => {
    expect(isWriteBlocked("PUT", "/workflows/wf_1/agents/a_1", VIEWER)).toBe(
      false,
    );
    expect(
      isWriteBlocked("POST", "/workflows/wf_1/agents/a_1/fund", VIEWER),
    ).toBe(false);
  });
});

File: backend/internal/api/readonly_test.go
package api

import (
	"net/http"
	"net/http/httptest"
	"testing"
)

func TestBlocksWrite(t *testing.T) {
	tests := []struct {
		name   string
		method string
		path   string
		want   bool
	}{
		// The five endpoints that change what a workflow is.
		{"create", http.MethodPost, "/workflows", true},
		{"update", http.MethodPut, "/workflows/wf_123", true},
		{"delete", http.MethodDelete, "/workflows/wf_123", true},
		{"deploy", http.MethodPost, "/workflows/wf_123/deploy", true},
		{"build", http.MethodPost, "/workflows/wf_123/build", true},

		// Schedule and geofence configuration change what triggers a workflow,
		// same category of authoring as the five above. Must mirror WRITE_RULES
		// in frontend/src/lib/readonly.ts one for one.
		{"set schedule", http.MethodPut, "/workflows/wf_123/schedule", true},
		{"clear schedule", http.MethodDelete, "/workflows/wf_123/schedule", true},
		{"set geofence", http.MethodPut, "/workflows/wf_123/geofence", true},
		{"clear geofence", http.MethodDelete, "/workflows/wf_123/geofence", true},
		// GET, but find-or-creates a workflow row server-side (see the comment
		// on readOnlyBlocked), so it belongs on this list despite the verb.
		{"tendril console", http.MethodGet, "/tendril/console", true},
		// The read-only counterpart never creates anything, so it stays open.
		{"tendril console exists", http.MethodGet, "/tendril/console/exists", false},
		// The Prism console's pair behaves identically, for the same reasons.
		{"prism console", http.MethodGet, "/prism/console", true},
		{"prism console exists", http.MethodGet, "/prism/console/exists", false},
		// Running a paid Prism call is operating, not authoring -- the same
		// line /tendril/run and /tendril/topup already sit on. A viewer who
		// can spend their own credit on a Tendril job can spend it here too.
		{"prism run", http.MethodPost, "/prism/run", false},
		{"prism endpoints", http.MethodGet, "/prism/endpoints", false},

		// Operating a workflow somebody else built stays available -- this is
		// the line the whole feature draws.
		{"run", http.MethodPost, "/workflows/wf_123/run", false},
		{"stop", http.MethodPost, "/workflows/wf_123/stop", false},
		{"list", http.MethodGet, "/workflows", false},
		{"get", http.MethodGet, "/workflows/wf_123", false},

		// Account and billing are not workflow authoring.
		{"settings", http.MethodPatch, "/settings", false},
		{"password", http.MethodPost, "/auth/password", false},
		{"coupon", http.MethodPost, "/credits/redeem-coupon", false},
		{"topup", http.MethodPost, "/tendril/topup", false},

		// A trailing slash routes to the same handler, so it must not be a
		// way around the rules.
		{"create trailing slash", http.MethodPost, "/workflows/", true},
		{"update trailing slash", http.MethodPut, "/workflows/wf_1/", true},

		// The id is exactly one path segment; anything deeper is a different
		// endpoint and must not be swept up.
		{"agent balance", http.MethodGet, "/workflows/wf_1/agents/a_1/balance", false},
		{"fund agent", http.MethodPost, "/workflows/wf_1/agents/a_1/fund", false},

		// A blocked path under a different verb is a different endpoint.
		{"get deploy", http.MethodGet, "/workflows/wf_1/deploy", false},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			if got := blocksWrite(tc.method, tc.path); got != tc.want {
				t.Fatalf("blocksWrite(%q, %q) = %v, want %v", tc.method, tc.path, got, tc.want)
			}
		})
	}
}

// ok is the handler the middleware wraps; a 200 means the request got through.
func ok(w http.ResponseWriter, _ *http.Request) { w.WriteHeader(http.StatusOK) }

func serve(t *testing.T, mw func(http.Handler) http.Handler, method, path, key string) int {
	t.Helper()
	req := httptest.NewRequest(method, path, nil)
	if key != "" {
		req.Header.Set(editorKeyHeaderKey, key)
	}
	rec := httptest.NewRecorder()
	mw(http.HandlerFunc(ok)).ServeHTTP(rec, req)
	return rec.Code
}

func TestReadOnlyMiddlewareDisabledByDefault(t *testing.T) {
	// No WEB_READONLY_MODE set: an existing deployment must keep working
	// exactly as it did.
	mw := NewReadOnlyMiddleware()
	if got := serve(t, mw, http.MethodPost, "/workflows", ""); got != http.StatusOK {
		t.Fatalf("create with read-only off = %d, want %d", got, http.StatusOK)
	}
}

func TestReadOnlyMiddlewareBlocksWrites(t *testing.T) {
	t.Setenv(readOnlyEnvVar, "1")
	mw := NewReadOnlyMiddleware()

	if got := serve(t, mw, http.MethodPost, "/workflows", ""); got != http.StatusForbidden {
		t.Fatalf("create = %d, want %d", got, http.StatusForbidden)
	}
	if got := serve(t, mw, http.MethodPost, "/workflows/wf_1/run", ""); got != http.StatusOK {
		t.Fatalf("run = %d, want %d", got, http.StatusOK)
	}
	if got := serve(t, mw, http.MethodGet, "/workflows", ""); got != http.StatusOK {
		t.Fatalf("list = %d, want %d", got, http.StatusOK)
	}
}

func TestReadOnlyMiddlewareEditorKey(t *testing.T) {
	t.Setenv(readOnlyEnvVar, "true")
	t.Setenv(editorKeyEnvVar, "s3cret")
	mw := NewReadOnlyMiddleware()

	if got := serve(t, mw, http.MethodPost, "/workflows", "s3cret"); got != http.StatusOK {
		t.Fatalf("create with editor key = %d, want %d", got, http.StatusOK)
	}
	if got := serve(t, mw, http.MethodPost, "/workflows", "wrong"); got != http.StatusForbidden {
		t.Fatalf("create with wrong key = %d, want %d", got, http.StatusForbidden)
	}
	if got := serve(t, mw, http.MethodPost, "/workflows", ""); got != http.StatusForbidden {
		t.Fatalf("create with no key = %d, want %d", got, http.StatusForbidden)
	}
}

// An unset EDITOR_CLIENT_KEY must not mean "every client is an editor" -- a
// missing header would otherwise compare equal to a missing key and let the
// whole blocklist through.
func TestReadOnlyMiddlewareEmptyEditorKeyGrantsNothing(t *testing.T) {
	t.Setenv(readOnlyEnvVar, "1")
	t.Setenv(editorKeyEnvVar, "")
	mw := NewReadOnlyMiddleware()

	if got := serve(t, mw, http.MethodPost, "/workflows", ""); got != http.StatusForbidden {
		t.Fatalf("no key configured, no header = %d, want %d", got, http.StatusForbidden)
	}
	if got := serve(t, mw, http.MethodPost, "/workflows", "anything"); got != http.StatusForbidden {
		t.Fatalf("no key configured, header sent = %d, want %d", got, http.StatusForbidden)
	}
}

func TestIsTruthy(t *testing.T) {
	for _, v := range []string{"1", "true", "TRUE", " yes ", "on"} {
		if !isTruthy(v) {
			t.Errorf("isTruthy(%q) = false, want true", v)
		}
	}
	for _, v := range []string{"", "0", "false", "no", "off", "maybe"} {
		if isTruthy(v) {
			t.Errorf("isTruthy(%q) = true, want false", v)
		}
	}
}
